Link Control: Clear entity metadata when selecting custom URLs

jeryj opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

Splitting out work from https://github.com/WordPress/gutenberg/pull/73731/

## What?
<!-- Link this PR to its associated issue with an appropriate keyword: Closes, See, Follow up to, etc. -->
Closes <!-- #ISSUE-NUMBER or URL --> https://github.com/WordPress/gutenberg/issues/73767, but I think there may be a further underlying issue there.

Fixes entity metadata (type/kind) persisting when changing from page links to custom URLs. 




<!-- In a few words, what is the PR actually doing? -->

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
Changing to from an entity-bound link to a custom link would persist the kind/type from the previously bound link, which causes odd behavior due to unexpectedly combined values.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
  - Refactor onChange from LinkControl into a hook with defensive checks for updating from an entity to custom link
  - LinkControl: Use LINK_ENTRY_TYPES to detect custom URL selections
  - LinkControl: Improve isEntity check to exclude custom URLs by type field
  - Add unit tests
  - Add e2e test

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->
1. Go to a navigation link item with a bound entity link
2. Edit the link via the link control by clicking the unsync button then typing a new url like "https://test.com"
3. Click Apply
4. The link should be updated to a custom link with value https://test.com

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

## Screenshots or screencast <!-- if applicable -->

<!-- If you would like to upload screenshots, feel free to use the table below when it is useful to show the difference between before and after the change. -->

|Before|After|
|-|-|
|<!-- Before screenshot here -->|<!-- After screenshot here -->|

jeryj committed:
Add fix for link control setting custom links as page links when updating a page link to a custom link

jeryj committed:
Link Control: Clear entity metadata when selecting custom URLs

jeryj committed:
Fix updating from bound entity to custom link

getdave commented on packages/block-editor/src/components/link-control/index.js:
```diff
@@ -207,8 +207,16 @@ function LinkControl( {
+	const isEntity =
+		handleEntities &&
+		!! internalControlValue?.id &&
+		! isCustomURLType( internalControlValue );
```
Ah so custom links had an `id` 🤦

jeryj replied:
In the unit tests, the custom links were getting set with an id value of the url.

getdave commented on packages/block-editor/src/components/link-control/index.js:
```diff
@@ -292,9 +300,15 @@ function LinkControl( {
 			...internalControlValue,
 			...nonSettingsChanges,
+			// Explicitly set type and kind to undefined for custom links
+			...( isCustomLink && { type: undefined, kind: undefined } ),
```
Just wondering if we should unset `id` here also?

jeryj replied:
I wasn't sure. I was hesitant to unset an ID since it previously had one. I figured I would let the ID do what it was before rather than introduce a potential breaking change.

getdave replied:
I don't like that the tests don't reflect the code. It could mean we are masking issues...

getdave reviewed (approved):
Thanks for following up on this.

I swapped from entity to static URL and saw

```
<!-- wp:navigation-link {"label":"Nihil ipsam cupiditate voluptates quidem","type":"custom","url":"www.wordpress.org","kind":"custom"} /-->
```

This looks correct to me. No bindings and nothing else hanging around.


I did notice a potential legacy issue with code and tests having different `type` values for custom links.

The tests seem to use 'URL' (uppercase) as the `type` but the actual code produces `type: link` (lowercase) via `URL_TYPE = 'link'` in constants. I wonder if this might be causing other issues?

jeryj commented:
> The tests seem to use 'URL' (uppercase) as the type but the actual code produces type: link (lowercase) via URL_TYPE = 'link' in constants. I wonder if this might be causing other issues?

I noticed this too. I wasn't sure if it was intentional to use 'URL' or not, but the tests pass with the new changes so I assumed it was fine.

jeryj committed:
LinkControl: Exclude kind/type from currentLink when selecting suggestions

Prevents old entity metadata from persisting when changing links. When a new suggestion is selected, only the new suggestion's kind and type are used, not the values from the previous link.

jeryj committed:
Set Id to undefined for custom links

jeryj committed:
Update tests from 'URL' to 'link'

jeryj commented:
@getdave I made a few more changes and wanted additional confidence before merging. Specifically [this commit](https://github.com/WordPress/gutenberg/pull/73825/commits/b2dd4f417dd619ac00e05384725dd6d572a07795). I believe this was a root bug that was passing the old `kind` and `type` from the entities when entering a custom link.

getdave commented on packages/block-editor/src/components/link-control/search-input.js:
```diff
@@ -106,7 +106,8 @@ const LinkControlSearchInput = forwardRef(
 				( suggestion && Object.keys( suggestion ).length >= 1 )
 			) {
-				const { id, url, ...restLinkProps } = currentLink ?? {};
+				const { id, url, kind, type, ...restLinkProps } =
```
Let's comment on what we're doing and why here.

jeryj replied:
Updated. This is actually the root bug fix. I'm going to just bring this change in, as the others are more heavy-handed about forcing the shape of custom url data and could be a breaking change if someone is expecting a certain return value.

getdave commented on packages/block-editor/src/components/link-control/test/index.js:
```diff
@@ -731,6 +731,33 @@ describe( 'Manual link entry', () => {
 		}
 	);
 
+	it( 'should return undefined kind and type when submitting a custom URL', async () => {
```
I'm thinking the test should be the control already has a link with kind and type and _then_ we change it and assert that the `kind` and `type` are no longer present.

jeryj replied:
That [test is later on](https://github.com/WordPress/gutenberg/pull/73825/files#diff-f133ef8bc484506e19d14be3b9f281b93b657b7fc472b757569e9b3548329483R2842): it( 'should clear entity metadata (type/kind) when changing from page link to custom link via suggestion',

getdave commented on packages/block-editor/src/components/link-control/index.js:
```diff
@@ -207,8 +207,16 @@ function LinkControl( {
 
+	// Helper to check if a link value has a custom URL type (not an entity type)
+	const isCustomURLType = ( linkValue ) =>
+		LINK_ENTRY_TYPES.includes( linkValue?.type );
```
@jeryj worth bringing this suggestion in?

jeryj replied:
No, the 'URL' was only in the unit tests. I updated that to be removed.

getdave commented on packages/block-editor/src/components/link-control/index.js:
```diff
@@ -207,8 +207,16 @@ function LinkControl( {
+
 	// Compute isEntity internally based on handleEntities prop and presence of ID
-	const isEntity = handleEntities && !! internalControlValue?.id;
+	// Exclude custom URLs which have id but type is 'link', 'mailto', 'tel', or 'internal'
```
```suggestion
	// Exclude custom URLs which (for historical reasons) may have 
	// an `id` but type is 'link', 'mailto', 'tel', or 'internal'
```

getdave commented on packages/block-editor/src/components/link-control/index.js:
```diff
@@ -292,9 +300,19 @@ function LinkControl( {
 		onChange( {
 			...internalControlValue,
 			...nonSettingsChanges,
+			// Explicitly set type and kind to undefined for custom links
```
```suggestion
			// Explicitly unset attributes for custom links to ensure consistency
			// and avoid edge cases.
```

getdave reviewed (approved):
This seems to be testing well for me.

I did notice that when using LinkControl in paragraph content we have a similar issue with attributes not being cleared on the link _format_ when a link changes from entity -> custom.

Could we apply a similar fix there?

https://github.com/WordPress/gutenberg/blob/27a57445cd870406dfaa4b07d351bff3f26aa5f6/packages/format-library/src/link/inline.js#L128-L138

The issue is nearly always when we merge the new attributes with the old attributes.

jeryj committed:
Improve and explain comments.

jeryj committed:
Undo link-control/index work, as it was on top of the root bug fix
